File: LC/NC250/valid_palindrome.c
// problem: https://leetcode.com/problems/valid-palindrome/

// Using 2 pointer approch
// time - O(N), space - O(1)
bool isPalindrome(char* s) {
  int l = 0, r = strlen(s) - 1;

  while (l < r) {
    // ignoring non alpha numeric characters
    while (l < r && !isalnum(s[l]) ) l++;
    while (l < r && !isalnum(s[r]) ) r--;

    if (tolower(s[l]) != tolower(s[r])) return false;

    l++, r--;
  }

  return true;    
}

File: LC/NC250/valid_palindrome.py
# problem: https://leetcode.com/problems/valid-palindrome/

# filter out the chars other than alpha numeric
# reverse the string and check for palindrome
# time - O(N), space - O(N)
class Solution:
  def isPalindrome(self, s: str) -> bool:
    # filter out other characters in the string
    filtered_str = "".join(filter(str.isalnum, s)).lower()

    return filtered_str == filtered_str[::-1]